### ISSUE TITLE
React edit forms still marked as dirty after closing them

### ISSUE THREAD

**Issue by GuySartorelli** [silverstripe/silverstripe-admin#2116]
In the assets section, if you start editing some file or folder and then navigate away (and select "OK" on the unsaved changes prompt), _every_ navigation within the admin will give you an unsaved changes prompt until you open another edit form (for any file or folder).
Basically the problem is that the form state is still marked "dirty" after navigating away, and that state is only changed when another (or the same) edit form is opened.

Observed in both CMS 4.12 and CMS 5.0 beta.

**Comment by GuySartorelli** [comment]
It's possible https://github.com/silverstripe/silverstripe-asset-admin/issues/1325 is the same problem, but it's hard to tell based on the information in that issue.

**Comment by GuySartorelli** [comment]
UPDATE: This seems to be a wider problem than just asset admin, stretching to all react sections. The same can be seen in campaign admin. Moving to silverstripe/admin.
I suspect this is related to the `NavigationBlocker` component.

**Comment by GuySartorelli** [comment]
Turns out this is _not_ the fault of the `NavigationBlocker` component, but is caused by setting `destroyOnUnmount: false` for redux form.
That was added in https://github.com/silverstripe/silverstripe-admin/pull/654 to support https://github.com/silverstripe/silverstripe-elemental/pull/397

tl;dr we don't remove form state from redux when a form is unmounted, because _maybe_ that form needs to restore the state on the next mount (though that's _only_ the case if the `refetchSchemaOnMount` prop is pass with `false` despite ALL forms remaining in redux state on unmount).

## Possible solutions

1. Go back to always destroying on unmount, and hide forms instead of unmounting them if they should be temporarily not displayed.
2. Add an `effect` that takes advantage of `unsavedFormsReducer` to remove forms from the "unsaved forms" state on unmount, and add them back again on mount if appropriate.
3. Only set `destroyOnMount` to `false` if the `refetchSchemaOnMount` prop is set to `false`.

All of these have problems:

1.  That's just revisiting the previous state of affairs again and presumably there was a good reason to move away from that. My guess is it'll break a bunch of stuff we then need to go and find, and fix.
2. While relatively straight forward, it does introduce a new race condition where navigation won't be blocked after the form is unmounted and before it's remounted. Notably if an elemental block is collapsed after changes have been made, navigation would no longer be blocked.
3. This would work great for most forms, but there may be some forms which have `refetchSchemaOnMount` set `false` but can't be easily reopened when navigation is blocked, i.e. this issue may still remain for those forms. That said I strongly suspect `refetchSchemaOnMount` is only ever used in elemental, and if that's the case this is definitely the best way forward IMO.
